### PR TITLE
feat: Add SpeculativeDecodingTrainer with TRAIN_ONLY and DATA_ONLY mode support

### PR DESCRIPTION
**Pull request by hrathina**
**What this PR does / why we need it**:

Adds `SpeculativeDecodingTrainer` to the Kubeflow SDK, enabling users to train Eagle3 speculative decoding draft models through two modes:

  ### TRAIN_ONLY mode
  Trains an Eagle3 draft model from **pre-extracted** hidden states and a preprocessed Arrow dataset. The user provides `hidden_states_path` and `data_path` pointing to existing data on a PVC. Training runs via `torchrun` with
  FSDP support for multi-GPU. This mode is for users who have already generated their training data (e.g., via a prior DATA_ONLY run) and want to iterate on training hyperparameters without re-extracting hidden states.
  
  ### DATA_ONLY mode
  Extracts hidden states from a verifier model using a **vLLM sidecar** container. The SDK automatically deploys a vLLM server alongside the training pod, configured with `extract_hidden_states` speculative decoding and the
  `ExampleHiddenStatesConnector` for writing hidden states to shared PVC storage. The user provides a `dataset_name` (built-in like `sharegpt`, a HuggingFace dataset ID, or a local JSONL file). Optionally supports
  `regenerate_responses` to re-generate dataset responses from the verifier model before extraction, and `max_samples` to limit the dataset size.

  ### Key capabilities
  - **PVC-based storage**: All paths (`output_dir`, `hidden_states_path`, `data_path`, `verifier_model`, `dataset_name`) support `pvc://` URIs with automatic volume mounting and path resolution
  - **Progression tracking**: In-memory HTTP metrics endpoint reports real-time training/extraction progress to the Kubeflow controller
  - **AutoConfig layer detection**: Automatically detects `target_layer_ids` from the verifier model's HuggingFace config (falls back to manual config for PVC-hosted models)
  - **vLLM sidecar management**: For DATA_ONLY, the SDK configures the sidecar with `speculative-config`, `kv-transfer-config`, GPU allocation, and shared PVC mounts
  - **Script injection**: Training and data generation scripts are embedded and injected into the training pod via `inspect.getsource()` — no external image dependencies beyond the base runtime

Fixes #

**Verification steps** _(required; provide valid, reproducible steps for reviewers to verify your changes, e.g. setup, commands, expected outcome, notebooks)_:

_Example: 1. Check out this branch. 2. Run `make test-python`. 3. Confirm that …_

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/) included if any changes are user facing

### CODE REVIEW

**Review by efazal — COMMENTED** [review]
Thanks @hrathina. The implementation looks great. I have few comments:

File: kubeflow/trainer/rhai/speculator.py [hunk 1]
```diff
@@ -0,0 +1,1490 @@
...
+        "norm_before_residual": norm_before_residual,
+        "norm_before_fc": norm_before_fc,
+        "embed_requires_grad": embed_requires_grad,
+        "ttt_steps": ttt_steps,
+        "verifier_name_or_path": verifier_model,
+        "d2t": d2t,
+        "t2d": t2d,
+    }
+    if from_pretrained is not None:
+        from_training_kwargs["from_pretrained"] = from_pretrained
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py]: `from_pretrained` is silently ignored here. Verified against speculators 0.6.0: `Eagle3DraftModel.from_training_args()` never reads this kwarg — it accesses specific named keys and discards the rest.

A user setting `SpeculatorConfig(from_pretrained="/path/to/ckpt")` will train from scratch with no error. Either implement resume logic after model creation, or raise `NotImplementedError` if this is set.

File: kubeflow/trainer/rhai/speculator.py [hunk 2]
```diff
@@ -0,0 +1,1490 @@
...
+    if target_layer_ids is not None:
+        from_training_kwargs["target_layer_ids"] = target_layer_ids
+    model = Eagle3DraftModel.from_training_args(verifier_config, **from_training_kwargs)
+
+    max_len = total_seq_len
+    collate_fn = create_collate_fn(max_len, verifier_config.hidden_size)
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py]: Missing `dtype` parameter. Full signature: `create_collate_fn(max_len, hidden_size, num_target_layers=3, dtype=torch.bfloat16, preprocess=None)`.

When user sets `hidden_states_dtype="float16"`, `ArrowDataset` loads float16 but collate creates bfloat16 padding. Dtype mismatch crash or silent precision loss.

Fix: `create_collate_fn(max_len, verifier_config.hidden_size, dtype=hs_dtype)`

File: kubeflow/trainer/rhai/speculator.py [hunk 3]
```diff
@@ -0,0 +1,1490 @@
...
+        spec_dict["initContainers"] = []
+
+    if trainer.verifier_model.startswith(PVC_URI_SCHEME):
+        resolved_verifier, _ = parse_output_dir_uri(trainer.verifier_model)
+    else:
+        resolved_verifier = trainer.verifier_model
+
+    cfg = trainer.config or SpeculatorConfig()
+
+    layer_ids_str = ",".join(str(lid) for lid in cfg.target_layer_ids)
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:1423]: Two issues with this env var:

1. **Format**: Produces comma-separated `"2,18,33"` but the CTR sidecar inserts this into a JSON arg for vLLM `--speculative-config`, which expects a JSON array `[2,18,33,36]`. See [vLLM hidden state extraction docs](https://docs.vllm.ai/en/latest/features/speculative_decoding/extract_hidden_states/).

2. **Missing last layer**: Only includes the 3 auxiliary layers. vLLM `eagle_aux_hidden_state_layer_ids` also needs `num_hidden_layers` appended for logit computation. The TrainingHub PR does this explicitly: `extraction_layer_ids = list(target_layer_ids); extraction_layer_ids.append(num_hidden_layers)`.

File: kubeflow/trainer/rhai/speculator.py [hunk 4]
```diff
@@ -0,0 +1,1490 @@
...
+    from speculators.train.noise_transforms import AddUniformNoise
+    from speculators.train.trainer import Trainer, TrainerConfig
+    from speculators.train.vocab_mapping import build_vocab_mappings_from_distribution
+    import torch
+    from torch.utils.data import DataLoader
+    from transformers import AutoConfig
+
+    rank = int(os.environ.get("RANK", "0"))
+    marker_name = f"{EXTRACTION_INCOMPLETE_MARKER}.rank-{rank}"  # noqa: F821
+    hs_dir = os.path.join(hidden_states_path, "hidden_states")
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py]: Double nesting. The parameter name `hidden_states_path` implies it points to the hidden states directory, but the code appends `/hidden_states` again.

If a user points this to where files actually are (from data_only output), the code looks at `<path>/hidden_states/hidden_states/` — which doesn't exist.

Either remove the appending, or rename the parameter and document that it should point to the parent directory.

File: kubeflow/trainer/rhai/speculator.py [hunk 5]
```diff
@@ -0,0 +1,1490 @@
...
+    else:
+        resolved_draft_vocab = draft_vocab_size or min(8192, target_vocab_size)
+        token_freq_path = Path(data_path) / "token_freq.pt"
+        token_freq_dict = torch.load(str(token_freq_path), weights_only=True)
+        d2t, t2d = build_vocab_mappings_from_distribution(
+            token_freq_dict=token_freq_dict,
+            draft_vocab_size=resolved_draft_vocab,
+            target_vocab_size=target_vocab_size,
+        )
+        np.save(str(d2t_path), d2t.cpu().numpy())
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:718]: Race condition in multi-GPU mode. All torchrun ranks hit the `else` branch simultaneously and write to the same `.npy` files concurrently. No rank guard or barrier exists.

Fix: only rank 0 should compute and save, then `dist.barrier()` before other ranks load.

File: kubeflow/trainer/rhai/speculator.py [hunk 6]
```diff
@@ -0,0 +1,1490 @@
...
+        )
+
+    local_rank = int(os.environ.get("LOCAL_RANK", 0))
+    is_distributed = int(os.environ.get("WORLD_SIZE", 1)) > 1
+    if is_distributed and not torch.distributed.is_initialized():
+        torch.distributed.init_process_group(backend="nccl")
+        torch.cuda.set_device(local_rank)
+
+    verifier_config = AutoConfig.from_pretrained(verifier_model)
+    target_vocab_size = verifier_config.vocab_size
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:704]: Missing `text_config` traversal. Some model architectures wrap the text config inside a parent config object. `AutoConfig` returns the wrapper where `vocab_size` may be missing or incorrect.

The `__post_init__` at line 404 correctly does:
```python
if hasattr(model_config, "text_config"):
    model_config = model_config.text_config
```

This function should do the same.

File: kubeflow/trainer/rhai/speculator.py [hunk 7]
```diff
@@ -0,0 +1,1490 @@
...
+    verifier_model: str
+    mode: SpeculatorMode
+    speculator_type: SpeculatorType = SpeculatorType.EAGLE3
+    hidden_states_path: str | None = None
+    data_path: str | None = None
+    dataset_name: str | None = None
+    max_samples: int | None = None
+    epochs: int = 3
+    lr: float = 1e-4
+    total_seq_len: int = 8192
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py]: The TrainingHub PR defaults `total_seq_len` to 2048 (`params.get("total_seq_len", 2048)`). A default of 8192 increases the risk of OOM for users who do not explicitly set this, since both the training process and the vLLM sidecar allocate GPU memory proportional to sequence length (KV cache scales linearly with context length).

File: kubeflow/trainer/rhai/speculator.py [hunk 8]
```diff
@@ -0,0 +1,1490 @@
...
+            flush=True,
+        )
+
+    if "_set_phase" in globals():
+        _set_phase("checking_vllm", 5)  # noqa: F821
+
+    endpoint = vllm_endpoint
+    print(f"[Kubeflow] Using vLLM endpoint: {endpoint}", flush=True)
+    health = vllm_endpoint.rstrip("/").rsplit("/v1", 1)[0] + "/health"
+    timeout_secs = 600
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py]: 600s may not be enough for larger verifier models that take significant time to load into GPU memory. Consider increasing to 1200s or making it configurable via a trainer parameter so users can adjust based on their model size and hardware.

File: kubeflow/trainer/rhai/scripts/response_regeneration.py [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+            for _ in range(len(workers)):
+                await queue.put(None)
+            await asyncio.gather(*workers)
+
+
+if __name__ == "__main__":
+    try:
+        asyncio.run(main())
+    except KeyboardInterrupt:
+        sys.exit(130)
```
Comment (efazal) [inline kubeflow/trainer/rhai/scripts/response_regeneration.py]: Missing trailing newline.
